File: Algorithm_235_Lowest Common Ancestor of a Binary Search Tree.cpp
#include <iostream>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
bool find(TreeNode *root, TreeNode *p)
{
	if (!root) return false;
	if (root==p) return true;
	return find(root->left,p) || find(root->right,p);
}
TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q)
{
	if (!root) return NULL;
	if (find(root->left,p) && find(root->left,q)) return lowestCommonAncestor(root->left,p,q);
	if (find(root->right,p) && find(root->right,q)) return lowestCommonAncestor(root->right,p,q);
	return root;
}
int main()
{
	TreeNode *t1=new TreeNode(1);
	TreeNode *t2=new TreeNode(2);
	TreeNode *t3=new TreeNode(3);
	TreeNode *t4=new TreeNode(4); 
	TreeNode *t5=new TreeNode(5);
	t1->left=t2,
	t1->right=t3;
	t2->left=t4;
	t2->right=t5;
	TreeNode *tmp=lowestCommonAncestor(t1,t4,t5);
	cout<<tmp->val;
	return 0;
}

File: Algorithm_235_Lowest Common Ancestor of a Binary Search Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// 方法1：简单方法。由于树是BST，可以通过比较根节点和p、q节点的值的大小
// 来判断应该向左子树还是右子树移动

public class SimpleSolution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root.val > p.val && root.val > q.val){
            return lowestCommonAncestor(root.left, p, q);
        }else if(root.val < p.val && root.val < q.val){
            return lowestCommonAncestor(root.right, p, q);
        }else{
            return root;
        }
    }
}

// 方法2：通用方法，类似236

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) {
            return null;
        } else if (root == p) {
            return p;
        } else if (root == q) {
            return q;
        }
        TreeNode l = lowestCommonAncestor(root.left, p, q);
        TreeNode r = lowestCommonAncestor(root.right, p, q);
        if (l != null && r != null) {
            return root;
        } else if (l != null) {
            return l;
        } else {
            return r;
        }
    }
}